File: backend/app/routers/staff.py
from datetime import date as date_type

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import StaffGroup, Staff, StaffSkill, StaffPermittedShift, Leave, SkillValue, Shift, User
from app.dependencies.auth import get_current_user
from app.schemas.staff import (
    StaffGroupCreate, StaffGroupUpdate, StaffGroupOut,
    StaffCreate, StaffUpdate, StaffOut,
    StaffSkillAdd, StaffPermittedShiftAdd,
    LeaveCreate, LeaveUpdate, LeaveOut,
)

router = APIRouter(prefix="/staff", tags=["Staff"])


# ── Staff Groups ─────────────────────────────────────────────────────

@router.get("/groups", response_model=list[StaffGroupOut])
def list_staff_groups(db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return db.query(StaffGroup).order_by(StaffGroup.name).all()


@router.post("/groups", response_model=StaffGroupOut, status_code=status.HTTP_201_CREATED)
def create_staff_group(body: StaffGroupCreate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if db.query(StaffGroup).filter_by(name=body.name).first():
        raise HTTPException(status.HTTP_409_CONFLICT, f"Staff group '{body.name}' already exists.")
    sg = StaffGroup(**body.model_dump())
    db.add(sg)
    db.commit()
    db.refresh(sg)
    return sg


@router.patch("/groups/{group_id}", response_model=StaffGroupOut)
def update_staff_group(group_id: int, body: StaffGroupUpdate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    sg = db.get(StaffGroup, group_id)
    if not sg:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff group not found.")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(sg, field, value)
    db.commit()
    db.refresh(sg)
    return sg


@router.delete("/groups/{group_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_staff_group(group_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    sg = db.get(StaffGroup, group_id)
    if not sg:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff group not found.")
    if sg.staff:
        raise HTTPException(status.HTTP_409_CONFLICT, "Cannot delete a group that still has staff members.")
    db.delete(sg)
    db.commit()


# ── Leaves ───────────────────────────────────────────────────────────
# NOTE: these must be defined BEFORE /{staff_id} routes to avoid
# FastAPI matching "leaves" as a staff_id integer parameter.

@router.get("/leaves", response_model=list[LeaveOut])
def list_leaves(
    staff_id: int | None = None,
    from_date: date_type | None = None,
    to_date: date_type | None = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = db.query(Leave)
    if staff_id:
        q = q.filter_by(staff_id=staff_id)
    if from_date:
        q = q.filter(Leave.date >= from_date)
    if to_date:
        q = q.filter(Leave.date <= to_date)
    return q.order_by(Leave.date).all()


@router.post("/leaves", response_model=LeaveOut, status_code=status.HTTP_201_CREATED)
def create_leave(body: LeaveCreate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not db.get(Staff, body.staff_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    leave = Leave(**body.model_dump())
    db.add(leave)
    db.commit()
    db.refresh(leave)
    return leave


@router.patch("/leaves/{leave_id}", response_model=LeaveOut)
def update_leave(leave_id: int, body: LeaveUpdate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    leave = db.get(Leave, leave_id)
    if not leave:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Leave not found.")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(leave, field, value)
    db.commit()
    db.refresh(leave)
    return leave


@router.delete("/leaves/{leave_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_leave(leave_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    leave = db.get(Leave, leave_id)
    if not leave:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Leave not found.")
    db.delete(leave)
    db.commit()


# ── Staff ────────────────────────────────────────────────────────────

@router.get("", response_model=list[StaffOut])
def list_staff(
    group_id: int | None = None,
    include_deleted: bool = Query(False),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = db.query(Staff)
    if not include_deleted:
        q = q.filter_by(deleted=False)
    if group_id:
        q = q.filter_by(staff_group_id=group_id)
    return q.order_by(Staff.full_name).all()


@router.post("", response_model=StaffOut, status_code=status.HTTP_201_CREATED)
def create_staff(body: StaffCreate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not db.get(StaffGroup, body.staff_group_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff group not found.")
    if db.query(Staff).filter_by(employee_id=body.employee_id).first():
        raise HTTPException(status.HTTP_409_CONFLICT, f"Employee ID '{body.employee_id}' already exists.")
    s = Staff(**body.model_dump())
    db.add(s)
    db.commit()
    db.refresh(s)
    return s


@router.get("/{staff_id}", response_model=StaffOut)
def get_staff(staff_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    return s


@router.patch("/{staff_id}", response_model=StaffOut)
def update_staff(staff_id: int, body: StaffUpdate, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(s, field, value)
    db.commit()
    db.refresh(s)
    return s


@router.post("/{staff_id}/delete", response_model=StaffOut)
def soft_delete_staff(staff_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Soft-delete: sets deleted=True. Staff remains in historical rosters."""
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    s.deleted = True
    db.commit()
    db.refresh(s)
    return s


@router.post("/{staff_id}/restore", response_model=StaffOut)
def restore_staff(staff_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    s.deleted = False
    db.commit()
    db.refresh(s)
    return s


# ── Staff Skills ─────────────────────────────────────────────────────

@router.get("/{staff_id}/skills")
def get_staff_skills(staff_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    return [
        {
            "skill_value_id": ss.skill_value_id,
            "skill_type":     ss.skill_value.skill_type.name,
            "value":          ss.skill_value.value,
        }
        for ss in s.skills
    ]


@router.post("/{staff_id}/skills", status_code=status.HTTP_201_CREATED)
def add_staff_skill(staff_id: int, body: StaffSkillAdd, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not db.get(Staff, staff_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    if not db.get(SkillValue, body.skill_value_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Skill value not found.")
    existing = db.query(StaffSkill).filter_by(staff_id=staff_id, skill_value_id=body.skill_value_id).first()
    if existing:
        raise HTTPException(status.HTTP_409_CONFLICT, "Skill already assigned to this staff member.")
    db.add(StaffSkill(staff_id=staff_id, skill_value_id=body.skill_value_id))
    db.commit()
    return {"staff_id": staff_id, "skill_value_id": body.skill_value_id}


@router.delete("/{staff_id}/skills/{skill_value_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_staff_skill(staff_id: int, skill_value_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    sk = db.query(StaffSkill).filter_by(staff_id=staff_id, skill_value_id=skill_value_id).first()
    if not sk:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Skill assignment not found.")
    db.delete(sk)
    db.commit()


# ── Permitted Shifts ─────────────────────────────────────────────────

@router.get("/{staff_id}/permitted-shifts")
def get_permitted_shifts(staff_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    s = db.get(Staff, staff_id)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    if not s.permitted_shifts:
        return {"note": "No restrictions — all shifts permitted", "permitted_shifts": []}
    return {
        "note": "Restricted to listed shifts only",
        "permitted_shifts": [
            {
                "shift_id":   ps.shift_id,
                "shift_code": ps.shift.code,
                "shift_name": ps.shift.name,
                "group":      ps.shift.group.code,
            }
            for ps in s.permitted_shifts
        ]
    }


@router.post("/{staff_id}/permitted-shifts", status_code=status.HTTP_201_CREATED)
def add_permitted_shift(staff_id: int, body: StaffPermittedShiftAdd, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not db.get(Staff, staff_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Staff not found.")
    if not db.get(Shift, body.shift_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Shift not found.")
    existing = db.query(StaffPermittedShift).filter_by(staff_id=staff_id, shift_id=body.shift_id).first()
    if existing:
        raise HTTPException(status.HTTP_409_CONFLICT, "Shift already permitted for this staff member.")
    db.add(StaffPermittedShift(staff_id=staff_id, shift_id=body.shift_id))
    db.commit()
    return {"staff_id": staff_id, "shift_id": body.shift_id}


@router.delete("/{staff_id}/permitted-shifts/{shift_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_permitted_shift(staff_id: int, shift_id: int, db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    ps = db.query(StaffPermittedShift).filter_by(staff_id=staff_id, shift_id=shift_id).first()
    if not ps:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Permitted shift not found.")
    db.delete(ps)
    db.commit()

File: frontend/src/api/staff.ts
import { api } from "./client";

export interface StaffGroup {
  id: number;
  name: string;
}

export interface StaffGroupInput {
  name: string;
}

export interface Staff {
  id: number;
  staff_group_id: number;
  employee_id: string;
  full_name: string;
  deleted: boolean;
  staff_group: StaffGroup;
}

export interface StaffInput {
  staff_group_id: number;
  employee_id: string;
  full_name: string;
}

export interface StaffSkill {
  skill_value_id: number;
  skill_type: string;
  value: string;
}

export interface PermittedShiftEntry {
  shift_id: number;
  shift_code: string;
  shift_name: string;
  group: string;
}

export interface PermittedShiftsResponse {
  note: string;
  permitted_shifts: PermittedShiftEntry[];
}

export interface Leave {
  id: number;
  staff_id: number;
  date: string;
  shift_code: string;
  note: string | null;
}

export interface LeaveInput {
  staff_id: number;
  date: string;
  shift_code?: string;
  note?: string | null;
}

// Staff groups
export const listStaffGroups = () =>
  api.get<StaffGroup[]>("/staff/groups").then((r) => r.data);

export const createStaffGroup = (body: StaffGroupInput) =>
  api.post<StaffGroup>("/staff/groups", body).then((r) => r.data);

export const updateStaffGroup = (id: number, body: Partial<StaffGroupInput>) =>
  api.patch<StaffGroup>(`/staff/groups/${id}`, body).then((r) => r.data);

export const deleteStaffGroup = (id: number) =>
  api.delete(`/staff/groups/${id}`).then(() => undefined);

// Staff
export const listStaff = (params?: { group_id?: number; include_deleted?: boolean }) =>
  api.get<Staff[]>("/staff", { params }).then((r) => r.data);

export const getStaff = (id: number) =>
  api.get<Staff>(`/staff/${id}`).then((r) => r.data);

export const createStaff = (body: StaffInput) =>
  api.post<Staff>("/staff", body).then((r) => r.data);

export const updateStaff = (id: number, body: Partial<StaffInput>) =>
  api.patch<Staff>(`/staff/${id}`, body).then((r) => r.data);

export const softDeleteStaff = (id: number) =>
  api.post<Staff>(`/staff/${id}/delete`).then((r) => r.data);

export const restoreStaff = (id: number) =>
  api.post<Staff>(`/staff/${id}/restore`).then((r) => r.data);

// Skills
export const getStaffSkills = (staff_id: number) =>
  api.get<StaffSkill[]>(`/staff/${staff_id}/skills`).then((r) => r.data);

export const addStaffSkill = (staff_id: number, skill_value_id: number) =>
  api.post(`/staff/${staff_id}/skills`, { skill_value_id }).then(() => undefined);

export const removeStaffSkill = (staff_id: number, skill_value_id: number) =>
  api.delete(`/staff/${staff_id}/skills/${skill_value_id}`).then(() => undefined);

// Permitted shifts
export const getPermittedShifts = (staff_id: number) =>
  api.get<PermittedShiftsResponse>(`/staff/${staff_id}/permitted-shifts`).then((r) => r.data);

export const addPermittedShift = (staff_id: number, shift_id: number) =>
  api.post(`/staff/${staff_id}/permitted-shifts`, { shift_id }).then(() => undefined);

export const removePermittedShift = (staff_id: number, shift_id: number) =>
  api.delete(`/staff/${staff_id}/permitted-shifts/${shift_id}`).then(() => undefined);

// Leaves
export const listLeaves = (params?: { staff_id?: number; from_date?: string; to_date?: string }) =>
  api.get<Leave[]>("/staff/leaves", { params }).then((r) => r.data);

export const createLeave = (body: LeaveInput) =>
  api.post<Leave>("/staff/leaves", body).then((r) => r.data);

export const updateLeave = (id: number, body: Partial<Pick<LeaveInput, "shift_code" | "note">>) =>
  api.patch<Leave>(`/staff/leaves/${id}`, body).then((r) => r.data);

export const deleteLeave = (id: number) =>
  api.delete(`/staff/leaves/${id}`).then(() => undefined);
